### ISSUE TITLE
No Revocation check on Certificates

### ISSUE THREAD

**Issue by iamsilvio** [Digitaler-Impfnachweis/covpass-ios#19]
Page 5 of [verifiable vaccination certificates - basic interoperability elements](https://ec.europa.eu/health/sites/default/files/ehealth/docs/vaccination-proof_interoperability-guidelines_en.pdf) states:

> A trust framework, including digital infrastructure, that is needed for establishing the 
authenticity and validity of certificates presented by certificate holders. 

Page 11 of [Interoperability of health certificates 
Trust framework](https://ec.europa.eu/health/sites/default/files/ehealth/docs/trust-framework_interoperability_certificates_en.pdf) states :

> 3.1.1.7 Health certificate revocation list (HCRL) 
A system used by Country A for publishing information about revoked health certificates.  
Each  Country  A  shall publish  one  and  only one  aggregate  list  of  all  revoked  health 
certificates. Country A is responsible for putting its revoked certificates on the list and signing 
it using one of its signing keys controlled by the PHA.

I do not see any implementation of a revocation check.

**Comment by oliver-steinbrecher** [comment]
HCRL is currently out of scope on european level

**Comment by iamsilvio** [comment]
Outside of which scope? 
Are there comprehensible reasons for this? 
Checking for crl is no magic, and it has a big impact on the confidentiality of the solution!

**Comment by oliver-steinbrecher** [comment]
The hcrl is meant to provide a list of individual certificates which have been revoked. This does not scale on national and European level. We’re able to revoke Document Signing Certificates (DSC) but such a revocation will effect all issued DCCs signed by that dsc.  The eu does not leverage the standard for crl instead they issue a list of trustworthy DSCs.

**Comment by martin-verlage** [comment]
Hm, the case in Friesland obviously shows the need for revoking issued documents. A nurse exchanged the doses and salt water was injected. The current solution smells like "fire and forget". In my understanding there is no way to invalidate the certificates issued by pharmacies, as neither the place where the injection was provided is in the certificate enclosed nor any data is hosted centrally. This is a significant drawback and far beyond what was specified in spring this year.

**Comment by oliver-steinbrecher** [comment]
There is an ongoing European discussion on how to address certificate revocations. We're going to adopt the solution once it has been clarified within the ehealth network.

**Comment by lennybacon** [comment]
As long as there is no solution implemented I do not see a reason why the issue was closed. The issue in the real world is till alive:

https://www.watson.ch/digital/schweiz/419566012-zertifikate-dealer-packt-aus-wir-versorgen-schweizer-kunden

**Comment by martin-verlage** [comment]
I fully support @lennybacon to leave the issue open - it is a serious issue. A further example [false certificates in Bavaria (sorry, in German language only)](https://www.br.de/nachrichten/bayern/impfbetrug-wemdinger-sollen-corona-schutz-ueberpruefen-lassen), shows the full consequences for holders: by press release the holders are requested to not longer use the original documents (z.B. gelber Impfpass) as well as the issued EU certificates, otherwise there is a risk of criminal penalties (Androhung strafrechtlicher Konsequenzen).

**Comment by Steuerboykott** [comment]
Anyone interested in hacking the EU Green Pass tyranny can get from me what I found on Telegram. I found private keys on Telegram, but I don't know exactly how to implement them to a working program that creates certificates. There was one project on Github that could create certificates in the correct format, but they were not recognized to be valid.
Send me your message by email and include your Github user name, so I can check if you are contributing something useful. If you are working on a web tool so everyone (!) can create their own and valid certificate from home, you have my sympathy.
Programmers: I will send you what I have and you then will create some certificates from names and birthdates given by me. If I have proof, then let's release to the public. My concern is interest-free if it's for the use of all.
My political standpoint is (all in German) here...

- Marcel Barz: Die Pandemie in den Rohdaten, https://www.youtube.com/watch?v=nEPiOEkkWzg
- Tom Lausen beim Exomagazin, https://www.youtube.com/watch?v=oqUtg6oPJBU
- Tom Lausen bei Punkt.PRERADOVIC, https://www.youtube.com/watch?v=0sBM7MVG8Lo
- Sebastian Götz: Die Zerstörung des Corona Hypes, https://www.youtube.com/watch?v=Juugv0T7inc
- Sebastian Götz: Haben wir Corona unterschätzt?, https://www.youtube.com/watch?v=rDBn1f8CSZI
- Paul Schreyer: Pandemie-Planspiele – Vorbereitung einer neuen Ära, https://www.youtube.com/watch?v=SSnJhHOU_28

**Comment by Ein-Tim** [comment]
@jleufgen

The last comment should probably be hidden for off topic or be deleted?

**Comment by Steuerboykott** [comment]
> @jleufgen
> 
> The last comment should probably be hidden for off topic or be deleted?

Of course not. We will get you all, be sure.

**Comment by RichiH** [comment]
https://docs.github.com/en/communities/maintaining-your-safety-on-github/reporting-abuse-or-spam#reporting-a-user gibt's nicht auf deutsch, aber funktioniert.

**Comment by Ein-Tim** [comment]
@RichiH 

Thank! I reported both comments. However, IHMO, they should still be deleted. 

The same thing was commented in https://github.com/ehn-dcc-development/hcert-spec/discussions/105 and it was deleted 2 min. later.

**Comment by Steuerboykott** [comment]
> 
> 
> @RichiH
> 
> Thanks! I reported both comments. However, IHMO, they should still be deleted.
> 
> The same thing was commented in [ehn-dcc-development/hcert-spec#105](https://github.com/ehn-dcc-development/hcert-spec/discussions/105) and it was deleted 2 min. later.

You are putting healthy people in quarantine while your faulty tests can not even differ between Covid and a regular flu. However it can not even prove an infection as several courts already have proven.
So you should be put behind bars. And you will be.

**Comment by oliver-steinbrecher** [comment]
I removed the content

**Comment by Ein-Tim** [comment]
Thanks @oliver-steinbrecher!

**Comment by iamsilvio** [comment]
> please do change the title of the issue to something more specific & concrete

The issue is still closed, what other title should cause this to be recognized as an issue?

Fire! Fire! Fire!

**Comment by lennybacon** [comment]
Today I was **still** able to use **MICKEY MOUSE** and other certificates

- to import it into the cwa on iOS - it shows and validates it as valid
- validate it with the CovPassCheck app - it shows and validates it as valid

So I'm asking POLITELY to reopen the issue **as this is a big f***in problem**.

![2021-11-10 13 11 21](https://user-images.githubusercontent.com/272164/141136170-4d2e57bd-4cd3-4b26-a5bc-a5a3e0fddb55.jpg)
 
![2021-11-10 15 50 29](https://user-images.githubusercontent.com/272164/141136191-cd179078-1839-418c-9042-f6ef9117ac5b.png)

**Comment by martin-verlage** [comment]
Definitely this issue is an issue and should be kept open until a satisfying solution is implemented. After some more studies of the process of issuing a certificate in Germany my observations are as follows:

- There is a split in the issuing process which produces no trustful documents. All 20.000+ pharmacies potentially issue digital certificates. The Robert Koch Institute (RKI) provides "just" a technical platform and by design trusts the input provided by the issuers (i.e., the holder of the Document Signing Certificates (DSC) and the responsible party for checking the validity of the data for the document / certificate are different). The real issuers (pharamcies) do not own an individual DSC. Hence, there is no other way instead declaring all digital certificates as invalid. - In my eyes this is the killer argument to all issuer-based revoking ideas.

- The solution is on document level only and is open for GIGO (garbage in, garbage out - greetings to Mickey Mouse and Johnny Walker). See the comment of @iamsilvio above (signed document versus certificate). Certificates are issued even when names do not exist and other personal data is wrong (e.g., typo at birth date). Daily experience shows that a cross check of the digital certificate with other documents (i.e., passport, id) is not the rule. Hence, the chance of using a digital certificate being recognized valid without detecting the mismatch between certified person and holder is very high.

- Moreover, there is an ongoing discussion whether faking certificates is illegal. So, no revoking, and no penalty. What should be the reason not to fake? Benefit with little risk attached was always a motivator for unwanted behavior.

Simply speaking, the whole process is potentially canning and signing garbage. Improving can and pen does not help.

**Comment by molk-ibm** [comment]
The dccs mentioned have not been issued by the german issuer. Currently, certs illegally issued in Germany will be rejected by CovPass Check after Nov. 15th, allowing legally issued certificate holders to re-issue their certificates.

There is a single pharmacy that issued valid certificates illegally. The only counter measure is to make CovPass Check to reject those certificates.

Rejecting the certificates mentioned (Mickey Mouse et. al.) can only be done on a EU level.

**Comment by oliver-steinbrecher** [comment]
The discussion about European wide revocation mechanisms are still ongoing. Multiple approaches are in discussion within the working group of the ehealth network. These details cannot be shared at the moment.

**Comment by lennybacon** [comment]
Dear project owners the communication, especially on this topic here, was non-transparent and the helpful and free volunteer work obviously not wanted. Keep going if you want the ones standing, helping and fighting for the same goals become frustrated until they turn of...

@molk-ibm of course it would be nice to have an EU wide solution. It would have been even more nice if anybody at EU level would have had brought up the idea that revocation is a necessity and cannot be postponed.I measure *can only be done on a EU level* as: We do not really care, will not build a tentative deny-list, and wait for somebody else fix it.

@oliver-steinbrecher it's a complete different story that the working group does something behind the closed curtains. I did not expect different - community work is not valued. Security researchers, especially in Germany, know this pretty well.
But sharing information was not what we asked for. **Our request was to re-open the issue until the problem, which you cannot deny actually exists, is fixed by whoever - just to be transparent about: There is an issue and somebody is working on it. **

**Comment by oliver-steinbrecher** [comment]
In order to provide more clarity please have a look into the published documents which describe the current agreed solution for DCC handling in the EU: 
- Overview: https://ec.europa.eu/health/ehealth/covid-19_en 
- Trustframework: https://ec.europa.eu/health/sites/health/files/ehealth/docs/trust-framework_interoperability_certificates_en.pdf

Please read them to have a better understanding.

The current trust framework is based on publishing CSCA and DSC lists which are considered as the anchor of trust for validation. CSCA and DSC revocations can be accomplished by removing them from the trust list.  This mechanism is used but does not allow individual revocation of one specific issued DCC.  This is going to be addressed in the next version of the trust framework. 

In Germany we have additional capabilities to revoke specific DCC based on the location-id in the UVCI. This allows us to revoke  DCCs issued by a specific pharmacy . In contrast to the EU framework this approach only works within Germany with the CovPass Check.  This mechanism is also in use.  

In summary it’s neither valid to say that we’re lacking revocation capabilities nor does the work on further trust framework improvements happen behind 'closed curtains'.

**Comment by martin-verlage** [comment]
@oliver-steinbrecher, I think it's not a god idea to refer to docs for a better understanding, when the docs contain exactly the demanded functionality which is not working (yet) - see section 3.1.1.7 HCRL in the trust framework outline. Moreover, as my most significant other is a pharmacist I can tell you that the location-id approach is causing massive confusion at the time of writing in this thread (see for example, https://www.deutsche-apotheker-zeitung.de/news/artikel/2021/11/16/alle-zertifikate-aus-einer-apotheke-fuer-ungueltig-erklaert). The location-id approach is a design bomb for the unlikely case that the issuer is the bad guy. The usual case of certificates which need revocation is that fake data is brought to the issuer, the issuer checks with high effort but does not detect the fake and issues the certificate, and later on, the fake is detected. This is far less than 1% of all certificates issued by that issuer, but still significant. Fighting the fake certificates with the location-id approach would also invalidate the other 99,+% of certificates issued by that issuer. Due to harming the issuers' business by using it (many regulars are affected), the location-id approach is very likely to be never used. The bulk of problems is not approached.

**Comment by oliver-steinbrecher** [comment]
The hcrl was not implemented due to privacy issues on European level. Hence the working group is discussing an alternative and enhanced approach to revoke specific DCCs. The solution Outline will be part of the next version of the related documents.

**Comment by lennybacon** [comment]
As I said before. There is no interest to a) inform anybody about ongoing discussions in (technical) committees on EU level b) discuss issues c) work on solutions.

This reminds me of the early Microsoft open source. It's just a placebo. Sad about the potential wasted. 

<sarcasm>
But hey, look at how well we cope with Covid-19. The politics get it all right and maybe in 3- 5 years we leave the Virus behind
</sarcasm>

**Comment by martin-verlage** [comment]
After drilling down into the subject a little bit my interpretation of the current status of implementation of revoking by location-id is as follows: a) the issuing body (i.e., a pharmacy) is represented in the system with a unique (?) identifier "location-id", b) the "location-id" is not a specific field in the DCC, c) the "location-id" is encoded in the "unique certificate identifier" (UCI), d) there are algorithms to retrieve the location-id from a given UCI, e) the "location-id based revokation mechanism" is an algorithm implemented in the CWA and CovPassCheck-App user device code declaring all certificates as illegal where the location-id matches a list of suspicious / non trustworthy / to-be-blocked location-ids, f) the revoking mechanism requires an update of the app, the to-be-blocked location-ids are not handled in a separate list, g) following the link in the above comment of @chscheiber the to-be-blocked location-ids are not available to developers of checking apps others than CovPass und CWA. @oliver-steinbrecher: can you confirm?

**Comment by molk-ibm** [comment]
@martin-verlage this is the way it is currently implemented, yes. as it is open source, you can check the implementation in the German apps yourself. The _algorithm_ is a simple list of hashes of the location ids to be blocked.
this national approach shall be replaced by a centralized one on eu-level, soon.
